### ISSUE TITLE
Config UI design a pipeline page that's compliant with the new `/pipelines` api

### ISSUE THREAD

**Issue by klesh** [merico-dev/lake#835]
## Description
When [Create an entity to represent a tasks Triggeration · Issue #824 · merico-dev/lake](https://github.com/merico-dev/lake/issues/824) done, `POST /task` behave differently. `config-ui` need a new design for task management to work properly

The coming pipeline api design is much more like gitlabs. In gitlab, a `pipeline` contains a set of `job`; in devlake, a `pipeline` contains a set of `task`, `job/task` share the same execution manner as following screenshot

![image](https://user-images.githubusercontent.com/61080/142788339-fa487da4-fe1d-4282-8af5-51db4cee7d81.png)

And users can manage them both `pipeline` and `task` based on what they need:
![image](https://user-images.githubusercontent.com/61080/142788498-9ef8b780-a495-4c38-8afb-4b56ff486d7d.png)

## Describe the solution you'd like
Design our `pipeline` and `task` management similar to `pipeline` and `job` from gitlab. no need to be as complete as `gitlab` but share the same idea to fulfill our need at the moment.

And I would like the `tasks` layout to be arranged accordingly to our json body as following:
![image](https://user-images.githubusercontent.com/61080/142795641-1cdb1be3-310e-49b6-9be1-e19163e93bf8.png)

## Additional context
This depends on issue #824 
@e2corporation let me know if you need more context

**Comment by e2corporation** [comment]
This should be moved to the next iteration after further discussion around context & needs. We still have an open-ended set of tickets for Triggers/Tasks re-design.

**Comment by hezyin** [comment]
@e2corporation This seems related to revamping the design of the trigger page. `Pipelines` page will replace `Tasks` page in the new design. I'm gonna keep it in iter5, but only for the design phase, we don't have to finish the implementation of the new triggers and pipelines page within this iter. cc @klesh @Startrekzky @joncodo

**Comment by e2corporation** [comment]
## Manage Pipelines
#### User Stories

- [ ] As a **Developer**, I want to see a list of _ALL_ Pipelines
- [ ] As a **Developer**, I want to see a list of _ACTIVE_ Pipelines 
- [ ] As a **Developer**, I want to see a list of _COMPLETE_ Pipelines 
- [ ] As a **Developer**, I want to see a list of _FAILED_ Pipelines 
- [ ] As a **Developer**, I want to be able to SEARCH Pipelines
    - [ ] Search by Pipeline Name
    - [ ] Search by Pipeline Status
    - [ ] Search by Date?
- [ ] As a **Developer**, I want to see/filter the _STATUS_ of a Pipeline
    - [ ] Active (Running)
    - [ ] Complete
    - [ ] Failed
- [ ] As a **Developer**, I want to see Percentage (%) PROGRESS of a Pipeline (ie 87% )
- [ ] As a **Developer**, I want to be able to **_STOP_** a running Pipeline
- [ ] As a **Developer**, I may want the ability to **STOP ALL** running Pipelines in the event of a race condition or failure.
- [ ] As a **Developer**, I want to be able to DELETE a _Stopped_ Pipeline
- [ ] As a **Developer**, I want to be able to EDIT/CONFIGURE a Pipeline
- [ ] As a **Developer**, I want to be able to RESTART a Running Pipeline
- [ ] As a **Developer**, I want to be able to RESTART a Failed Pipeline
- [ ] As a **Developer**, I want to see overview details for a Pipeline
    - [ ] Pipeline Name
    - [ ] Last Run Date
    - [ ] Status
- [ ] As a **Developer**, I want to be able to click a Pipeline to see _Detailed Pipeline Activity_
- [ ] As an **Engineering Manager**, I want to see enabled **Data Providers** for each Pipeline
- [ ] As an **Engineering Manager**, I want to manage _ALL_ Pipelines
- [ ] As an **Engineering Manager**, I may want to be notified  of Pipeline FAILURES (via e-mail)
- [ ] As a **Developer**, I want to access to Pipelines API Documentation
- [ ] As a **Developer**, I want access to **Create New Pipeline** (Convenience Link)


## Pipeline Run Details
#### User Stories

- [ ] As a **Developer**, I want to see Start Date/Time of a Pipeline Run
- [ ] As a **Developer**, I want to see detailed Pipeline Run Activity
- [ ] As a **Developer**, I want to see the overall Pipeline _Status_
- [ ] As a **Developer**, I want to see the _Stages_ for a Pipeline
- [ ] As a **Developer**, I want to see the _Progress_ for each Stage
- [ ] As a **Developer**, I want to see the _Tasks_ for each stage
- [ ] As a **Developer**, I want to see the  Pipeline _Duration_ (Execution Time)
- [ ] As a **Developer**, I want to be able to RESTART a Running Pipeline
- [ ] As a **Developer**, I want to be able to RESTART a Failed Pipeline
- [ ] As a **Developer**, I want to be able to **Go BACK** to _Manage Pipelines_
- [ ] As a **Developer**, I want to see the Failure/Error Responses for a Stage
- [ ] As a **Developer**, I want to see the Failure/Error Responses for a Task
- [ ] As an **Engineering Manager**, I want to see a pipeline's health and performance.


## Create Pipeline (Blueprint)
#### User Stories

- [ ] As a **Developer**, I want to be able to CREATE a Pipeline
- [ ] As a **Developer**, I want to be able to EDIT a Pipeline
- [ ] As a **Developer**, I want to be able to DELETE a Pipeline
- [ ] As a **Developer**, I want to be able to enable/disable (choose) which Data Providers to configure for a pipeline.
- [ ] As a **Developer**, I want to be able to specify Trigger SETTINGS for a Data Provider
- [ ] As a **Developer**, I want to be able to have access to MANAGE Settings for a Data Provider (convenience link)
- [ ] As a **Developer** I may want to enable SCHEDULING for a Pipeline to trigger at a specified interval
- [ ] As a **Developer** I may want to have the option to RUN the pipeline once automatically after saving/creating a new pipeline.


@joncodo @hezyin @klesh @kevin-kline @Startrekzky

**Comment by e2corporation** [comment]
![create-pipeline](https://user-images.githubusercontent.com/1742233/144903120-9e43bb92-f729-44d6-9aec-6a5eccc5c67a.png)
![manage-pipelines](https://user-images.githubusercontent.com/1742233/144903129-e3d0762c-f5e2-4f6b-adb5-e29a970bb939.png)
![pipeline-activity](https://user-images.githubusercontent.com/1742233/144903136-755c5432-89a6-4b32-8699-7a8c009ec69f.png)

@joncodo @hezyin

**Comment by klesh** [comment]
Why is a `blueprint` page here?
Aren't we attack the `blueprint` feature on another issue?

**Comment by e2corporation** [comment]
> Why is a `blueprint` page here?
> Aren't we attack the `blueprint` feature on another issue?

@klesh I was asked to create these 3 concepts with the "blueprint" plan applied. Maybe we'll have to design an alternate concept without blueprint/plan so it matches implementation.

**Comment by Startrekzky** [comment]
Hi @e2corporation , thanks for the user stories and prototype.
Overall, I think 

- the information structure is unclear. It looks like the `Pipeline Blueprint` and `Pipeline Run` are mixed under the same tabs.
- there is too much text, operations and colors on theses pages that I don't know how to get started and finish the tasks.

..., and my feedback is focusing on these aspects.

1. There isn't a status for 'Pipeline Blueprint' object, as it's only a schedule, not a specific job. Therefore, there shouldn't be a 'status filtering menu' on this page. Same goes for the 'status' and 'duration' columns in the list.
![image](https://user-images.githubusercontent.com/14050754/145201469-65d4a355-eaeb-4d0d-9582-0d2b70ee02fb.png)
![image](https://user-images.githubusercontent.com/14050754/145198256-3c8dbdbb-5d4c-452b-987a-7703416375a7.png)
2. There should be a page for 'Pipeline Runs' to present their status, duration, and running time. Something like this:
![image](https://user-images.githubusercontent.com/14050754/145202421-8d3baf14-df66-401a-bce2-3a41a0116af0.png)

These two points above are the most important.
3. Can we remove the 'preview' section from the 'Manage Pipeline'? Since we already have it on the 'pipeline run activity/detail page'.
![image](https://user-images.githubusercontent.com/14050754/145193497-30fbb0d5-40fd-4574-a21d-d2f138a6983c.png)
2. 
3. Can the 'New' button be bigger with background color? This button should be one of the most catchy elements on the 'pipeline run list' page, as it's a user's very first action when he/she entered this page.
![image](https://user-images.githubusercontent.com/14050754/145199140-a4b8d77e-54e0-4d38-84fb-0e977a51e38b.png)

**Comment by e2corporation** [comment]
@Startrekzky As discussed in earlier meetings today, many of the discrepancies are due the fact that we are designing with the Blueprint(Plan) schema in mind however the current implementation isn't ready for the features demonstrated above. With our new direction we are trimming all of this down to support what Backend has designed, which is to primarily support Pipeline Runs for this coming iteration. 

>1. There isn't a status for 'Pipeline Blueprint' object, as it's only a schedule, not a specific job. Therefore, there shouldn't be a 'status filtering menu' on this page. Same goes for the 'status' and 'duration' columns in the list.

Since the pipeline runs would have a child relationship to the Blueprint, we could consolidate all this data into one "manage" screen which is what this mockup is providing a glimpse of. So the filtering would be filtering pipelines that have runs matching a specific status. The overall status of the pipeline is a reflection of the last run, so that's a bit of reasoning why this concept can be possible. A fourth "All Jobs" Page can still be design and be useful, it just wasn't requested however for these initial mockups.

> 2. There should be a page for 'Pipeline Runs' to present their status, duration, and running time. Something like this:

This is being accomplished by the "Pipelines Activity" screen above, which is a detailed look at the Pipeline (Blueprint) that would show the Run history.

> 3. What do users see on 'Pipeline Blueprint' page by default? A blank 'Pipeline blueprint' list? If so, can the 'New' button be bigger with background color? This button should be one of the most catchy elements on the 'pipeline run list' page, as it's a user's very first action when he/she entered this page.

In a no-data/first-time-use scenario a message would be displayed that there are no blueprints defined and that the user should create one to proceed. The "Add" functionality isn't accentuated in these mockups because we already agreed on what the "Create Pipeline" screen should look like, these mockups were more aimed at managing the pipelines. I do agree with you that the "Add New" button should stand out more and it will get a primary color accent during implementation.

> 4. What will users see after creating a pipeline. Jump back to the 'Pipeline Blueprint' list? And how does a user trigger a 'Pipeline Run', I'd like to see more pages to reflect a specific use case of 'collect data'.
Point 1&2 are the most important. Point 3&4 are of interaction-level and are also very important. Other minor issues such as aesthetic-level issues can be polished later.

After creating a Pipeline, we could automatically redirect the user to the "Pipeline Activity" page OR the "Manage Pipelines" page. Both can and should be capable of presenting the user with the status of the current active/running pipeline so they can easily see what pipeline is active. As far as the "Job Details" for a particular run/execution, that would consist of the Collection Steps that are now present on the triggers page.

**Comment by e2corporation** [comment]
Latest design mockups attached for our next discussion/questions round regarding **Pipelines**. Overall I think we are headed in the right direction towards supporting a minimal interface that will work with the current BE design. Looking forward to team feedback and next steps.

![pipelines-concepts-v1 1](https://user-images.githubusercontent.com/1742233/145660845-b38674be-cf53-4e2d-ab7e-da276c056b13.png)

@Startrekzky @joncodo @hezyin @kevin-kline @klesh

**Comment by klesh** [comment]
Great job, some minor wording:
1. Missed pipeline run name, we do have name for pipeline run.
2. Rename "stop" to "cancel" to algin BE/FE terminology
Logging windows is great, but keep in mind we dont have logging api at the moment.

**Comment by e2corporation** [comment]
**Pipelines Concepts V1.2**
Apply feedback from last review session on 12/13/2021.
- Add Support for **Pipeline Run Name** (Auto-generated default)
- Rename "Stop" to "CANCEL"
- Optimize Data-table controls for All Pipelines (Manage)
- Show Run Settings on Activity/Details page
- Text/Copy Updates

![pipelines-concepts-v1 2](https://user-images.githubusercontent.com/1742233/145911703-a9829fa1-8c4e-48f8-98a3-724d7a07ce47.png)

**Comment by Startrekzky** [comment]
Moved to #962
